game/src/contrib/utils/components/ai/transition/ProtectOnAttack.java

```diff
@@ -43,9 +43,15 @@ public ProtectOnAttack(Entity entity) {
      *
      * @param entities - Entities that are protected
      */
-    public ProtectOnAttack(Collection<Entity> entities) {
+    public ProtectOnAttack(final Collection<Entity> entities) {
         entities.stream()
-                .peek(e -> e.getComponent(HealthComponent.class).orElseThrow())
+                .peek(
```

Comment:
Ich verstehe den Sinn von diesem `peek` nicht: Du beschwerst Dich hier im Stream (!), wenn eine Entität keine `HealthComponent` hat - d.h. eigentlich möchtest Du eine Filter-Operation, die nur noch Entitäten mit `HealthComponent` übrig lässt für die Verarbeitung in `forEach`?

Reply:
Musste mich grade nochmal selbst in die Klasse einlesen:
Der Methode übergibst du eine Collection an Entitäten die du als Monster beschützt. 
Das bedeutet, wenn der Spieler eines dieser Monster haut, dann haust du den Spieler.
Dafür brauchen die zu beschützenden Entitäten ein HealthComponent. 
So das ist das drum herum.

Ein filter würde doch jetzt alle Entitäten "rauswerfen" die das nicht haben, aber keinen Fehler werfen, oder (klar könnte man da jetzt logging einbauen)? 
Heißt aber auch, ich als Kunde kann dir ungültige Entitäten übergeben.

Ist jetzt eine Geschmacksfrage, lieber Crashen oder lieber weiterspielen und ignorieren.

Reply:
#751

Reply:
Hmmm, das wirft für mich die neue Frage auf, warum ein Monster nur Entitäten mit HC beschützen kann? Kann es nicht auch eine Kiste beschützen?

Ich finde den `peek` nicht so gut lesbar an der Stelle (deshalb meine Nachfrage) - und traditionell ist das eher zum Debuggen gedacht wg. der Nebenwirkungen. Ich glaube, ich hätte dann lieber nach "hat keine HC" gefiltert und mit `count()` abgeschlossen und die Exception außerhalb der Stream-Operation geworfen, wenn das Ergebnis nicht leer ist. Danach könnte man das `add()` machen.

Also irgendwie so (kompiliert nicht, weil mir eine der "kaputten" Entitäten fehlt): 

```java
        long hasHC = entities.stream()
            .flatMap(e -> e.fetch(HealthComponent.class).stream())
            .count();
        if (hasHC < entities.size()) throw new MissingComponentException.build(???, HealthComponent.class));

        toProtect.addAll(entities);
```

Oder sowas:

```java
        List<MissingComponentException> missingHC = entities.stream()
            .filter(e -> e.fetch(HealthComponent.class).isEmpty())
            .map(e -> MissingComponentException.build(e, HealthComponent.class))
            .collect(Collectors.toList());
        if (missingHC.size() > 0) throw missingHC.get(0);

        toProtect.addAll(entities);
```